pkg/kubectl/secret_for_docker_registry.go

```diff
@@ -0,0 +1,131 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/credentialprovider"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+// SecretForDockerRegistryGeneratorV1 supports stable generation of a docker registry secret
+type SecretForDockerRegistryGeneratorV1 struct {
+	// Name of secret (required)
+	Name string
+	// Username for registry (required)
+	Username string
+	// Email for registry (required)
+	Email string
+	// Password for registry (required)
+	Password string
+	// Server for registry (required)
+	Server string
+}
+
+// Ensure it supports the generator pattern that uses parameter injection
+var _ Generator = &SecretForDockerRegistryGeneratorV1{}
+
+// Ensure it supports the generator pattern that uses parameters specified during construction
+var _ StructuredGenerator = &SecretForDockerRegistryGeneratorV1{}
+
+// Generate returns a secret using the specified parameters
+func (s SecretForDockerRegistryGeneratorV1) Generate(genericParams map[string]interface{}) (runtime.Object, error) {
+	err := ValidateParams(s.ParamNames(), genericParams)
+	if err != nil {
+		return nil, err
+	}
+	params := map[string]string{}
+	for key, value := range genericParams {
+		strVal, isString := value.(string)
+		if !isString {
+			return nil, fmt.Errorf("expected string, saw %v for '%s'", value, key)
+		}
+		params[key] = strVal
+	}
+	delegate := &SecretForDockerRegistryGeneratorV1{
+		Name:     params["name"],
+		Username: params["docker-username"],
+		Email:    params["docker-email"],
+		Password: params["docker-password"],
+		Server:   params["docker-server"],
+	}
+	return delegate.StructuredGenerate()
+}
+
+// StructuredGenerate outputs a secret object using the configured fields
+func (s SecretForDockerRegistryGeneratorV1) StructuredGenerate() (runtime.Object, error) {
+	if err := s.validate(); err != nil {
+		return nil, err
+	}
+	dockercfgContent, err := handleDockercfgContent(s.Username, s.Password, s.Email, s.Server)
+	if err != nil {
+		return nil, err
+	}
+	secret := &api.Secret{}
+	secret.Name = s.Name
+	secret.Type = api.SecretTypeDockercfg
+	secret.Data = map[string][]byte{}
+	secret.Data[api.DockerConfigKey] = dockercfgContent
+	return secret, nil
+}
+
+// ParamNames returns the set of supported input parameters when using the parameter injection generator pattern
+func (s SecretForDockerRegistryGeneratorV1) ParamNames() []GeneratorParam {
+	return []GeneratorParam{
+		{"name", true},
+		{"docker-username", true},
+		{"docker-email", true},
+		{"docker-password", true},
+		{"docker-server", true},
+	}
+}
+
+// validate validates required fields are set to support structured generation
+func (s SecretForDockerRegistryGeneratorV1) validate() error {
```

Comment:
The rationale was that `StructuredGenerator` was from the caller's perspective, and caller's did not have to call `Validate`.  A complaint I had with the current generator framework was that the proper place to call `Validate()` was ill-defined.  I thought not including it here made it implicit that the `StructuredGenerate()` did all requisite validation.